mlir/lib/Dialect/Rock/Transforms/TransformToMemref.cpp

```diff
@@ -0,0 +1,132 @@
+//===- TransformToMemref - MLIR rock.transform conversion pass ---===//
+//
+// Copyright 2020 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================
+//
+// This pass converts rock.blockwise_* ops to rock.threadwise_*
+// and lowers other higher-level ops like transform and fill in preparation for
+// the threadwise lowering
+//
+//===-----------------------------------------------------===//
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/Tuning/GeneralGemmBlockStructure.h"
+#include "mlir/Dialect/Rock/utility/builderUtils.h"
+#include "mlir/Dialect/Rock/utility/loweringUtils.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Rock/IR/MfmaInsnGroup.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Debug.h"
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKTRANSFORMTOMEMREFPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-transform-to-memref"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace mlir::rock;
+
+namespace {
+struct RockTransformToMemrefPass
+    : public rock::impl::RockTransformToMemrefPassBase<
+          RockTransformToMemrefPass> {
+  void runOnOperation() override;
+};
+
+//===----------------------------------------------------------------------===//
+// TransformOp conversion to MemRef
+//===----------------------------------------------------------------------===//
+struct TransformRewritePattern : public OpRewritePattern<TransformOp> {
+  using OpRewritePattern<TransformOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(TransformOp op,
+                                PatternRewriter &b) const override {
+    Location loc = op.getLoc();
+    auto srcType = op.getOperand().getType().cast<ShapedType>();
```

Comment:
Does this end up needing the cast? I could've sworn that if you use more specific methods you'll get a `TypedValue<ShapedType>`